File: Kotlin/src/kotlin/binary_search/binary_search.kt
package jobspls.binary_search

fun <T : Comparable<T>> List<T>.binary_search(find: T): Int? {
    var lower = 0
    var upper = this.size - 1
    var index: Int = 0

    while (lower <= upper) {
        index = (lower + upper).floorDiv(2)
        val atIndex = this[index]

        when {
            atIndex < find -> lower = index + 1
            atIndex > find -> upper = index - 1
            else -> return index
        }
    }

    index
}

File: TypeScript/src/binary_search/binary_search.ts
export interface SearchResult {
  // Index if found
  found_idx?: number;
  // Location near where the element should be if appropriately placed
  missing_idx?: number;
}

export default function binary_search<T>(vec: Array<T>, find: T): SearchResult {
  let lower = 0;
  let upper = vec.length - 1;
  let index = 0;

  while (lower <= upper) {
    index = Math.floor((lower + upper) / 2);

    if (vec[index] < find) {
      lower = index + 1;
    } else if (vec[index] > find) {
      upper = index - 1;
    } else {
      return { found_idx: index };
    }
  }

  return { missing_idx: index };
}
